### PR TITLE
feature(gwk): add FedMOON algo

### PR DESCRIPTION
**Pull request by KyeGuo**

### CODE REVIEW

File: fling/component/client/fedmoon_client.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        r"""
+        Overview:
+            Training step. The loss of fedmoon should be added to the original loss.
+        """
+        batch_x, batch_y = batch_data['x'], batch_data['y']
+        z, o = self.model(batch_x)
+        main_loss = criterion(o, batch_y)
+        # Calculate fedmoon loss.
+        cos = nn.CosineSimilarity(dim=-1)
+        z_glob, _ = self.glob_model(batch_x)
```
Comment (kxzxvbk) [inline fling/component/client/fedmoon_client.py]: For each forward calculation for ``self.glob_model`` and ``self.prev_model``, use ``with torch.no_grad()``.

File: fling/component/client/fedmoon_client.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        z_i = cos(z, z_glob)
+        logits = z_i.reshape(-1, 1)
+        for prev_model in self.prev_models:
```
Comment (kxzxvbk) [inline fling/component/client/fedmoon_client.py:63]: I don't think it is a good idea to forward calculate all ``self.prev_models``. As the training round is getting larger, ``self.prev_models`` will finally have infinite items. I think a queue with fixed length is better.

File: fling/component/client/fedmoon_client.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    def train(self, lr, device=None):
+        r"""
+        Overview:
+            Training function. The global model and prev model should be stored.
+        """
+        self._store_global_model(self.model)
+        mean_monitor_variables = super(FedMOONClient, self).train(lr=lr, device=device)
+        # Reset the global model to save memory.
+        self.glob_model = None
```
Comment (kxzxvbk) [inline fling/component/client/fedmoon_client.py]: Use ``del self.glob_model``

File: fling/model/moon_models.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+from torch import nn
```
Comment (kxzxvbk) [inline fling/model/moon_models.py]: Don't create a new model class. Reuse the previous resnet. For example, you can add a new argument ``mode`` to the previous  forward function. Modes can contain: ['compute-logit', 'compute-feature-logit'], and the default value is 'compute-logit'

File: fling/component/server/fedmoon_server.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+from typing import Dict
```
Comment (kxzxvbk) [inline fling/component/server/fedmoon_server.py]: This file is not needed. Reuse the base server after you modify the model.

File: fling/component/client/fedmoon_client.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        # The variable to store the previous models.
+        self.prev_models = []
+
+    def _store_prev_model(self, model: nn.Module) -> None:
+        r"""
+        Overview:
+            Store the prev model for fedmoon loss calculation.
+        """
+        self.prev_models.append(copy.deepcopy(model.to(self.device)))
```
Comment (kxzxvbk) [inline fling/component/client/fedmoon_client.py]: Why put to ``self.device``？

File: fling/component/client/fedmoon_client.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        # Add the main loss and fedmoon loss together.
+        loss = main_loss + self.mu * fedmoon_loss
+
+        y_pred = torch.argmax(o, dim=-1)
+
+        monitor.append(
+            {
+                'train_acc': torch.mean((y_pred == batch_y).float()).item(),
+                'main_loss': main_loss.item(),
+                'fedmoon_loss': fedmoon_loss.item(),
```
Comment (kxzxvbk) [inline fling/component/client/fedmoon_client.py]: Instead of recording fedmoon loss, please record ``self.mu * fedmoon_loss.item()``

**Comment by kxzxvbk** [comment]
Reformat the files by: ``source format.sh ./``

**Comment by kxzxvbk** [comment]
Add information about fedmoon in README.md

File: fling/model/cnn.py [hunk 8]
```diff
@@ -32,10 +33,26 @@ def __init__(
...
+            self.mlp = []
+            self.mlp.append(nn.Flatten())
+            self.mlp.append(nn.Linear(hidden_dims[-1], linear_hidden_dims[0]))
+            for i in range(len(linear_hidden_dims) - 1):
+                self.mlp.append(get_activation(name=activation))
+                self.mlp.append(nn.Linear(linear_hidden_dims[i], linear_hidden_dims[i + 1]))
+            self.mlp = nn.Sequential(*self.mlp)
+            self.fc = nn.Linear(linear_hidden_dims[-1], class_number)
+        else:
+            self.mlp = None
```
Comment (kxzxvbk) [inline fling/model/cnn.py]: Don't use ``self.mlp = None``, change it to ``self.mlp = nn.Identity()``

File: fling/model/cnn.py [hunk 9]
```diff
@@ -32,10 +33,26 @@ def __init__(
...
         x = self.layers(x)
         x = self.glp(x)
-        x = self.fc(x)
-        return x
+        if self.mlp:
+            x = self.mlp(x)
+        y = self.fc(x)
+        if mode == 'compute-feature-logit':
+            return x, y
+        else:
```
Comment (kxzxvbk) [inline fling/model/cnn.py:58]: List all possible cases like 'compute-feature-logit' or 'compute-logit'. Add an exception if no legal mode is provided.

File: fling/model/resnet.py [hunk 10]
```diff
@@ -176,7 +177,19 @@ def __init__(
...
+            self.mlp = []
+            self.mlp.append(nn.Flatten())
+            self.mlp.append(nn.Linear(features[len(layers) - 1] * block.expansion, linear_hidden_dims[0]))
+            for i in range(len(linear_hidden_dims) - 1):
+                self.mlp.append(nn.ReLU(inplace=True))
+                self.mlp.append(nn.Linear(linear_hidden_dims[i], linear_hidden_dims[i + 1]))
+            self.mlp = nn.Sequential(*self.mlp)
+            self.fc = nn.Linear(linear_hidden_dims[-1], class_number)
+        else:
+            self.mlp = None
```
Comment (kxzxvbk) [inline fling/model/resnet.py]: ``nn.Identity()``

File: fling/model/resnet.py [hunk 11]
```diff
@@ -231,7 +244,7 @@ def _make_layer(
 
         return nn.Sequential(*layers)
 
-    def _forward_impl(self, x: Tensor) -> Tensor:
+    def _forward_impl(self, x: Tensor, mode='compute-logit') -> Tensor:
```
Comment (kxzxvbk) [inline fling/model/resnet.py]: Typing lint: e.g. ``mode: str = 'compute-logit'``

File: argzoo/cifar100/cifar100_fedmoon_resnet_config.py [hunk 12]
```diff
@@ -18,9 +18,9 @@
         queue_len=1,
     ),
     model=dict(
-        name='resnet50',
+        name='resnet8',
         input_channel=3,
-        linear_hidden_dims=[1000, 1000, 256],
+        # linear_hidden_dims=[256],
```
Comment (kxzxvbk) [inline argzoo/cifar100/cifar100_fedmoon_resnet_config.py:23]: Remove this commented code.
